main.go

```diff
@@ -14,23 +15,405 @@ import (
+	if err != nil {
+		return cli.Exit(err, 1), "Unable to get consumer UUID"
+	}
+	var successMsg string
+
+	_, _, _, isColorful := getColorPreferences(ctx)
+
+	if uuid == "" {
+		username := ctx.String("username")
+		password := ctx.String("password")
+		organization := ctx.String("organization")
+		activationKeys := ctx.StringSlice("activation-key")
+
+		if len(activationKeys) == 0 {
+			if username == "" {
+				password = ""
+				scanner := bufio.NewScanner(os.Stdin)
+				fmt.Print("Username: ")
+				_ = scanner.Scan()
+				username = strings.TrimSpace(scanner.Text())
+			}
+			if password == "" {
+				fmt.Print("Password: ")
+				data, err := term.ReadPassword(int(os.Stdin.Fd()))
+				if err != nil {
+					return cli.Exit(err, 1), "Unable to read password"
+				}
+				password = string(data)
+				fmt.Printf("\n\n")
+			}
+		}
+
+		var s *spinner.Spinner
+		if isColorful {
+			s = spinner.New(spinner.CharSets[9], 100*time.Millisecond)
+			s.Suffix = " Connecting to Red Hat Subscription Management..."
+			s.Start()
+			defer s.Stop()
+		}
+
+		var err error
+		if len(activationKeys) > 0 {
+			err = registerActivationKey(
+				organization,
+				ctx.StringSlice("activation-key"),
+				ctx.String("server"))
+		} else {
+			if organization != "" {
+				err = registerPassword(username, password, organization, ctx.String("server"))
+			} else {
+				/* TODO: When organization was not specified using CLI option --organization and it is
+				   required, because user is member of more than one organization, then ask for
+				   the organization. RHSM D-Bus API can provide list of available organizations
+				   in this case. */
+				err = registerPassword(username, password, "", ctx.String("server"))
+			}
+		}
+		if err != nil {
+			return cli.Exit(err, 1), "Unable to register system to RHSM"
+		}
+		successMsg = "Connected to Red Hat Subscription Management"
+	} else {
+		successMsg = "This system is already connected to Red Hat Subscription Management"
+	}
+	return nil, successMsg
+}
+
+// connectAction tries to register system against Red Hat Subscription Management,
+// connect system to Red Hat Insights and it also tries to start rhcd service
+func connectAction(ctx *cli.Context) error {
+	var start time.Time
+	durations := make(map[string]time.Duration)
+	errorMessages := make(map[string]error)
+	hostname, err := os.Hostname()
+	if err != nil {
+		return cli.Exit(err, 1)
+	}
+
+	fmt.Printf("Connecting %v to %v.\nThis might take a few seconds.\n\n", hostname, Provider)
+
+	connectedPrefix, _, errorPrefix, isColorful := getColorPreferences(ctx)
+
+	/* 1. Register to RHSM, because we need to get consumer certificate. This blocks following action */
+	start = time.Now()
+	var returnedMsg string
+	err, returnedMsg = registerRHSM(ctx)
+	if err != nil {
+		errorMessages["rhsm"] = fmt.Errorf("cannot connect to Red Hat Subscription Management: %w", err)
+		fmt.Printf(errorPrefix + " Cannot connect to Red Hat Subscription Management\n")
```

Comment:
BTW: When this action is not successful, then it would be better to skip following steps ("Connecting to Red Hat Insights" and "Starting rhcd"), because these steps could not be successful anyway.

Reply:
Yea, that's a good point. This action is required for the subsequent two. Maybe that's reason enough to make it "different" than the other two. It must succeed; if it does not, the program should exit.

Reply:
Yeah, I agree. Done.